### ISSUE TITLE
Introduce hyper parameters and config

### ISSUE THREAD

**Issue by dmpetrov** [iterative/dvc#3393]
For an ML experiment, it is important to know metrics as well as the parameters that were used in order to get the metrics. Today there is no training/processing parameter concept in DVC which creates a problem when a user needs to visualize an experiment for example in some UI.

A common workaround is to track parameters as metrics. However, the meaning of metrics is different. All the UI tools (including `dvc metrics diff`) need to show deltas where deltas do not make sense to some types of params. For example, delta for learning rate might be ok to see (values are still better), but delta for a number of layers (32, 64 or 128) does not make sense, the same for not numeric params like strings.

Also, config/parameters are a pre-requisite for experiment management (#2799 or CI/CD scenarios) when DVC (or other automation tools) need to change training regarding provided parameters.

Another benefit of the "understanding" parameter - DVC can use this information during `repro`. For example, DVC can realize that a step `process` which depends on config file `config.json` should not be run despite the config file change because the metrics it uses were not changed.

We need to introduce the experiment config file/parameters file with a fixed structure that DVC can understand.

Open questions:
1. Filename. `config.json`, `dvcconfig.json`, `params.json`.
2. File format: json, text config/ini (name=value), Hydra, ... We can run a survey.
3. How to track param dependency for stages. We can introduce a new type of dependency: param. If it is given then the stage depends on the file and on particular params values. Like `dvc run -p learning_rate -p LL5_levels ...`.
4. DVC should probably support groups of params. Param name pattern could be used : `dvc run -p 'processing.*' ...`

**Comment by elgehelge** [comment]
Great initiative @dmpetrov. What do you imagine the files contain? Is it stuff like `required=true` or is it types like `learning_rate=float` or is it default values like `learning_rate=3.5e-5`?

**Comment by elgehelge** [comment]
I think putting parameters in a file might be a step in the wrong direction, but I need to think a little more 🤔

**Comment by dmpetrov** [comment]
@elgehelge yes, it is about config with options that you can read from your code.

Having a config file that many ML pipeline stages use (and depend on) is a quite usual approach. We just need to specify a subset of params instead.

I'd love to hear other options!

**Comment by elgehelge** [comment]
Okay. A config/params file could be a good first step, but not having parameters as commandline arguments would make day-to-day development a bit cumbersome: edit file, `run repro`, edit file, `run repro`, and so on... , and also it would be hard to plug external tool for running jobs into this.

I do like the idea of having a file of default parameters though. But the important word here is _default_, since it would still be nice to be able to overwrite them using the commandline, but maybe this is another issue that comes later 😅

1. **File name:** Naming. `config.json` might conflict with other project related configs. And any name containing the word "config" makes me think that it contains settings. I am all for `params.json`.
2. **File format:** First I thought `json`. But then I saw that `ini` files have groups, which I find super useful. For instance, at a previous job we hardcoded 3 different sets of default parameters into the code, one for training a tiny model (for debugging purposes), one for training a model for a small segment of clients, and one for training a full model. It could also be that maybe for external compute or CI you would like to train for longer, so you could have a params specifically for this. And I guess there are plenty more use-cases where you might want more than one set of defaults. For this reason a like `ini`, however, two very strong reasons _not_ to pick `ini`: (1) There is no formal standard, (2) There is no types, (3) Arrays er annoying to define. So, while researching on this I found `TOML`, which seems to solve all the problems related to `ini` (https://en.wikipedia.org/wiki/TOML)
3. Yes! 
4. I love it.

Even though commandline argument support might be a different issue, let's still talk about how that would be added. Maybe the dvc file could look something like this maybe??:
```
md5: fd75c91bb60baf9bf36e4f64f0c1c7e0
cmd: python train_model.py --lr ${learning_rate}
deps:
- md5: 53184cb9a5fcf12de2460a1545ea50d4
  path: train_model.py
- md5: 809776eaeef854443d39e4f76bd64c84.dir
  path: data/split
- json: '0.0005'
  param: learning_rate
outs:
- md5: 47b3781954c760e1bd19ac9e56e141b1
  path: model/mymodel.pkl
```
And you could have different steps depending on other parameters like `test_set_size`, and then run `dvc repro --learning_rate 0.0005 --test_set_size 0.2`

Together with "the build cache" (https://github.com/iterative/dvc/issues/1234) I think this would work quite well.

**Comment by elgehelge** [comment]
While taking a step in this direction, we might also want to consider to go all the way and add environment  variables as a dependency. Maybe like so?
```
deps:
- env: TF_ADJUST_SATURATION_FUSED
  value: 1
```
It is not something I fin important myself, not might be worth considering.

**Comment by dmpetrov** [comment]
@elgehelge thank you for the feedback. All very helpful.

1. Filename. Agree that `config` is too broad. `params` or `dvcparams` are better options.
2. ini and json are definitely the most common cases. We actually can support both: `params.conf` or `params.json` (error if both are presented).

Note, for automation scenarios and visualization, we need a command to show all the params in a unified way like:
```
$ dvc params
learning_rate = 0.005
num_classes = 1000
augument.VerticalFlip = True
augument.Rotate = [20, 60, 90]
process.UseOpenCL = True

# The actual params file content:
$ cat params.conf
learning_rate = 0.005
num_classes = 1000

[augument]
VerticalFlip = True
Rotate = [20, 60, 90]

[process]
UseOpenCL = True
```

@elgehelge also, there is some misunderstanding in `dvc run/repro`. We need a way of defining params in `dvc run` and changing params in `dvc repro`.

Examples: 
| Action  | Command |
| ------------- | ------------- |
| Define dependency to param in default params file  `params.conf`  | `dvc run -p learning_rate -d train.py -d images/ -o model.pkl python train.py`  |
| Define dependency to a param in a custome params file `my_param.conf` |  `dvc run -P my_param.conf -p learning_rate -d train.py -d images/ -o model.pkl python train.py` |
| Change a param in current workspace and repro | `dvc repro model.pkl.dvc -p learning_rate 0.004` |
| Change a param in branch, repro and commit | `dvc repro model.pkl.dvc -p learning_rate 0.004 --branch new_lr --commit-message "try new learning rate 0.004"` |

Note, in both `dvc repro` examples the command depends on the param file. However, it should NOT retrain the model if the params file changed without the corresponded param change.

PS: sorry for being too detailed. I'm trying to provide as many details as possible to make sure everybody on the same page.

**Comment by elgehelge** [comment]
@dmpetrov Great examples. Two comments.

* Is there really a need for the "Change a param in branch, repro and commit"-action? Would it be nicer to just make a new branch, do `dvc repro model.pkl.dvc -p learning_rate 0.004`, and then commit? In a real world scenario I think the most common flow would be a bit different: Imagine that you just used the default learning rate, and now you want to tune it. First, you add support for a new parameter in your code, and then you add the parameter as a dependency. So this connot be done in a single command anyways.

* Maybe I was not being clear in my previous comment. I think it is important to consider the use-case of having multiple "parameters sets". One example could be this. Let's say you have a model that takes a whole day to train, but you would still like a CI to test that the code can run. It that case we would like a parameter set for a small fast model, and a parameter set for the real deployable one. Using `toml` this could look like:
```
[bigmodel]
learning_rate = 0.005
epochs = 25
data.Languages = ["english", "spanish", "chinese"]
data.MaxSamples = -1

[testmodel]
learning_rate = 0.005
epochs = 2
data.Languages = ["english"]
data.MaxSamples = 1000
```
And using `json`, you would need to have two different files, and thus two different stages. I am not saying one is better than the other, we just have to discuss it.

There is probably many other use-cases of having multiple parameter sets? Like having multiple customers which require different model settings (maybe the data needs different preprocessing).

**Comment by dmpetrov** [comment]
> Would it be nicer to just make a new branch, do `dvc repro model.pkl.dvc -p learning_rate 0.004`, and then commit? .... First, you add support for a new parameter in your code, and then you add the parameter as a dependency. So, this cannot be done in a single command anyways.

@elgehelge, yes, this is the scenario "Change a param in current workspace and repro". You are right that this is a default mode for brand new parameters or functionality.

While the last scenario "Change a param in branch, repro and commit" is automation on top of the previous one. It is needed when a parameter was already extracted and you do hyper param search in an automated or semi-automated way - from a script for example or through some service.


> I think it is important to consider the use-case of having multiple "parameters sets"

Very good point! It was not mentioned in the description. Happy to discuss this.

My thoughts... In the current proposal, the params could be easily overwritten `dvc repro -p learning_rate 0.004`. This is not enough for the use case you brought. To support your use case we can overwrite the entire param file like `dvc repro --param-file params_test.conf -p learning_rate 0.004` while the default `dvc repro` will work with `params_conf.conf` which was defined in `dvc run --param-file params_prod.conf` or a default one `params.conf`.

@elgehelge what do you think about this proposal?
1. Is it ok to keep separate files, not separate sections? What is your preference? 
2. Do we need a default `params.conf` or we should force users to specify it in each `dvc run`?
3. Any other ideas?

**Comment by elgehelge** [comment]
1. Yes I think multiple files will support most use-casen including future ones this we might not be able to imagine now.
2. I think a default is nice. Most projects will probability only need one file, so the tool is easier to get started with if there is a standardized place to put your parameters.

I wonder if I could try helping out and get the code to know a little. Maybe I would need some pointers about where to start 😅

**Comment by dmpetrov** [comment]
@elgehelge sure! It would be great if you can take this issue.
This is a bit heavy issue. Let's chat quickly - I'll give you a bit more context around the process and point you to the right people who can help. Please ping me in Discord to schedule the chat (or shoot an email to dmitry@iterative.ai).

**Comment by casperdcl** [comment]
Regarding `conf`/`ini`-vs-`json`, I'd say if wanting any level of manual user-editing `yaml` is best :)

**Comment by elgehelge** [comment]
> Regarding conf/ini-vs-json, I'd say if wanting any level of manual user-editing yaml is best :)
Yes, `yaml` is an option too. I don't have enough experience with these different formats, so I don't have an opinion. One argument could also be that the `.dvc` file is `yaml` and it might be good with some consistency.

> `dvc repro model.pkl.dvc -p learning_rate 0.004`
What about data types in this cmd? It is going to be a problem with 4 and "4"?

**Comment by casperdcl** [comment]
I assume some sort of intelligent `try: float(value) except: value` type thing would be nice.

**Comment by casperdcl** [comment]
Also what about this format?

```yaml
deps:
- md5: 53184cb9a5fcf12de2460a1545ea50d4
  path: train_model.py
- md5: 809776eaeef854443d39e4f76bd64c84.dir
  path: data/split
- params:
  - learning_rate: 0.0005
  - filters: [32, 32, 64, 128, 64, 32, 1]
  - dropout: false
  - activation: "relu"
```

**Comment by elgehelge** [comment]
I talked with @efiop on Discord. We were actually talking about
```
deps:
  - param: myparam
    value: paramvalue
```
However the suggestion given by @casperdcl is more readable, and also it would allow `params.yaml` to look similar to the .dvc file.

**Comment by efiop** [comment]
I guess the main question here is how to specify the params configs that we need to read. Combining our suggestions it might look something like:
```
deps:
  - path: params.conf
    params:
    - learning_rate: 0.0005
    - filters: [32, 32, 64, 128, 64, 32, 1]
    - dropout: false
    - activation: "relu"
```
which, actually, looks pretty good to me (but then again I'm not a DS).  @dmpetrov @elgehelge how does it look to you guys?

**Comment by casperdcl** [comment]
starting to look a lot like https://github.com/iterative/dvc/issues/3439#issuecomment-594863859

```yaml
deps:
  - path: params.conf
    md5: 1q2w3e4r5t6y7u8i9o0p9o8i7u6y5t4r
    filter: python extract_params_default.py --learning-rate 0.0005 --filters '32, 32, 64, 128, 64, 32, 1' --dropout False --activation relu
```

**Comment by dmpetrov** [comment]
Just to make sure we are all on the same page.

1. We introduce a config file that can be used in many stages and a stage triggers repro only if a specific subset of the params was changed.
2. The config file format can be config/ini, json or yaml. The major question is - that's the default name and format.
3. Users can redefine the config file name for a particular stage (or each of the stages).
2. The subset of the params that a stage use should be specified in `.dvc` file like
```
eps:
  - path: params.conf
    params:
    - learning_rate: 0.0005
    - filters: [32, 32, 64, 128, 64, 32, 1]
    - dropout: false
    - activation: "relu"
```

Is anything need to be changed in these statements?

**Comment by dmpetrov** [comment]
> What about data types in this cmd? It is going to be a problem with 3 and "3"?

Good question. Do we need types for params at all? We can say all are strings because there is no reason to compare them in some advanced way, in contrast to metrics.

@elgehelge Do you have a good scenario in mind when param type is important and can give us more value?

**Comment by dmpetrov** [comment]
> starting to look a lot like [#3439 (comment)](https://github.com/iterative/dvc/issues/3439#issuecomment-594863859)

🤔 you are right, @casperdcl. It looks like another type of granular\conditional dependency.

**Comment by elgehelge** [comment]
Okay. Where to start. There is multiple different things I would like to achieve, and they kind of relates.

1. *Convenience*. Instead of having your preprocessing and model parameters spread over many files and hardcoded into your code, it would be convenient to be able to collect them one place, but in a way that does not chance all of your stages if just a single parameter is changed. Today multiple parameter files (one for each state) would be possible, but its a little less convenient. As @casperdcl point out this is indeed related to #3439 since this related to making a dependency to a subset of a file. Since we are no longer talking about what I would call "parameterised pipelines", I like the idea of not introducing a new `-p` option, but instead just extending the `-d` option to something like `-d somefile.yaml::learning_rate`

However, I am still envisioning "parameterised pipelines" which would help os achieve:

2. *Hyperparameter* tuning. Sometimes I would like to add a new parameter and try several values of that parameter. It would be nice if the actual values of the parameters are included in the .dvc file for reproducibility, but I would like to not have to change any files during the process. I think this requires command line arguments to be involved. This is about changing the actual parameters _at runtime_, and thus related to...

3. *Automation tools*. We would like automation tools to be able to eg. run two different versions of your pipeline (eg. one model for phones and one for a web app). Or maybe you have a tool for starting jobs where you would like to start many jobs with various parameters, maybe you are using a tool for automatic hyperparameter search.

Would the following run of commands make sense in your world?? Maybe "pipeline arguments" is a better name than "parameters"?
`dvc run -p augment.rotate=False -d prepare_data.py -d images/ -o prep_data.csv python prepare_data.py --rotate {augment.rotate}`
`dvc run -p learning_rate=0.004 -d train.py -d prep_data.csv -o model.pkl python train.py --lr {learning_rate}`
`dvc repro model.pkl.dvc -p learning_rate=0.01 -p augment.rotate=True`

**Comment by andronovhopf** [comment]
Just jumping in with a quick question- how would the parameters stored in the `param` file relate to what is hard-coded in, say, my tensorflow model to be trained? I can't imagine you'd want to automate the extraction of this meta-data from the code (since you'd have to build a meta-data extractor for each different modeling framework); would these parameters necessarily be arguments to the script that defines and trains your model?

**Comment by casperdcl** [comment]
I don't think hardcoded params should have anything to do with dvc. If I understand correctly @andronovhopf by "hardcoded params" you really mean constant code vars, which by definition aren't "parameters."

If we want to expose one of these vars (as in @elgehelge's last suggestion via CLI flags) then they become params.

I also suggest:

1. *implementing hyperparams*: keeping this issue to implementing `-p <param>=<value>` which gets saved in `<output>.dvc`
2. *implementing config*: moving a shared `config.yml|ini|json` discussion to a different issue (perhaps not even required if https://github.com/iterative/dvc/issues/3439#issuecomment-594863859 is implemented)
    + overriding default config with a different "config patch" file, using subsets of config files etc should also be moved to a separate issue

**Comment by andronovhopf** [comment]
@casperdcl so to make sure I understand... 
In @elgehelge's suggestion:
```
dvc run -p learning_rate=0.004 -d train.py -d prep_data.csv -o model.pkl python train.py --lr {learning_rate} ```

the `train.py` takes `{learning_rate}` as an argument. Generally, in order for DVC to track a hyperparameter of an experiment, that parameter must be "exposed" by being made into an argument to a process in the DVC pipeline.

Do I have that right?

**Comment by casperdcl** [comment]
@andronovhopf I think the syntax would be:

```
dvc run -p lr=0.001 -p activation=relu -o output script {lr} --act {activation}
```
which would yield `output.dvc` with:

```yaml
md5: 1q2w3e4r5t6y7u8i9o0p
cmd: script {lr}
deps:
- params:
  - lr: 0.001
  - activation: relu
outs:
- md5: 0p9o8i7u6y5t4r3e2w1q
  path: output
```

and `dvc repro -p activation=tanh output` would trigger the substitution of `{}`:

```
script 0.001 --act tanh
```

It would be up to the end user to implement their `script` to accept CLI args/params in the format they specified in the initial `dvc run` command...

**Comment by andronovhopf** [comment]
Thanks @casperdcl, think I get it now!

**Comment by casperdcl** [comment]
Based on everyone's comments I doubt anyone gets anything. At least I don't :D

**Comment by casperdcl** [comment]
@dmpetrov I thing the whole **Trigger for granular\conditional dependency** might be interesting to keep here...

Basically if I do

```bash
dvc run -p lr=0.001 -p activation=relu -o output script {lr} --act {activation}
dvc repro -p activation=tanh output  # second experient
dvc repro -p activation=sigmoid output
dvc repro -p activation=tanh output  # should not re-run, should checkout `output` from "second experiment"
```

This is what I was referring to with:

> keeping track of when these values have been changed and then later reverted back to their original values may require an internal config file inside the `.dvc/` dir...

**Comment by elgehelge** [comment]
> # should not re-run, should checkout `output` from "second experiment"

This would be solved by https://github.com/iterative/dvc/issues/1234 or am I wrong?

**Comment by elgehelge** [comment]
Two ways the output could look like
```yaml
md5: 1q2w3e4r5t6y7u8i9o0p
cmd: script {lr} --act {activation}
deps:
- params:
  - lr: 0.001
  - activation: relu
```
```yaml
md5: 1q2w3e4r5t6y7u8i9o0p
cmd: script 0.001 --act relu
```
The last one would not even require a new dependency to be implemented.

**Comment by efiop** [comment]
Guys, I think the discussion took a detour. E.g. `dvc run -p param=value` that was suggested assumes that dvc has some way of passing the param value to the command that it is running, which is a pretty strong assumption, for which I don't have a nice solution. The suggested `{paramname}` substitution is not that simple and we have a separate ticket for that. I would start with a much less strong assumption that your command will simply read the param conf file and will get from it what it needs.

So from my point of view, I would do the first step as follows:

support `-p conffile:paramname` for `dvc run`, which will open `conffile`, read `paramname` param and will save it to dvc-file as:
```
deps:
- path: conffile
  params:
  - paramname: paramvalue
```
and your command will be assumed to read `conffile` by itself, no need for `dvc` to pass it anything in any way.

For "default" conf file name, we could imply that if `-p paramname` is supplied, we try to look for it in default config files and if we don't find it - error out.

**Comment by dmpetrov** [comment]
Thank you @efiop! We have to keep the scope of this issue manageable and do not include the parameter passing (`python train.py {lr}`) as well as build/run-cache (#1234) into it. Both of these features are very useful but let's move step by step.

Also, I'd like to mention a related issue https://github.com/iterative/dvc/issues/3439#issuecomment-598595045 which might affect the format of saving params in dvc-files. CC @elgehelge

**Comment by casperdcl** [comment]
@efiop but your suggestion would imply that DVC would also have to edit the `conffile` itself, otherwise the `cmd` would never be aware of the params.

**Comment by dmpetrov** [comment]
@casperdcl could you please elaborate?

In this approach, it's user's responsibility to read params from the file (in `cmd` code) and set right dependencies in the stage (`dvc run`). So, the `cmd` (from `dvc run`) is aware of the params. The exception is `dvc repro -p learning_rate 0.004` which modifies the param file.

**Comment by casperdcl** [comment]
Exactly. But I wouldn't call it an "exception" - surely `repro` is the main reason for the existence of `run`?

**Comment by elgehelge** [comment]
> The exception is `dvc repro -p learning_rate 0.004` which modifies the param file.

Wow. I never thought that we would write to the params file. I think that's a bad idea. As a user I would like to run my pipeline with various configurations, but not change the state / make side-effects. Also I would like to depend on a parameter that might not have a  location. If I am forced to pointing to a location I will end up with dummy values like learning_rate=0

**Comment by dmpetrov** [comment]
> Wow. I never thought that we would write to the params file. I think that's a bad idea.

In the basic scenario we should not write to the params file. This is needed to **automate `repro`** with params. The automation scenarios where described in https://github.com/iterative/dvc/issues/3393#issuecomment-591152649 but let me rephrase the repro ones:
1. **[Manual]** User modifies param file manually and runs a regular `dvc repro`
2. **[Auto**] User runs `dvc repro -p learning_rate=0.03` where DVC modifies the file and runs repro
3. **[Auto with commit**] The previous command can be run with commit `-branch new_lr --commit-message "try new learning rate 0.03”`

**The current issue is only about the basic scenario (1).** We will create new issue(s) for the automation once the first iteration (pull request) is ready.

**Comment by elgehelge** [comment]
To me it feel weird to tell an automation tool to "run train_model.py from commit a1b2c3d4 with learning_rate=0.004", and then the tool will checkout that commit, _edit som files_, produce a model, and then shut down (the file changes will never appear anywhere - and should not). It feel weird that the params needs a file location. Anyways, that is a longer for fundamental discussion that we should take somewhere else. For now I will just follow orders 👨‍✈️

.. So in regards to the discussion in #3439, we would like `dvc run -p my_param` to create a dvc file that looks something like this?
```
deps:
  - md5: 500f914dd06cf1f588b8e281c8fb3dd8
    path: params.conf
    filter: python dvc.param_getter.py
    params:
    - my_param: paramvaluefoundinfile
```
??
Or could you help me out here?

**Comment by efiop** [comment]
@elgehelge I might be a bit lost in conversations mixed with #3439 , but it seems like we should not bother with `filter` in the first iteration. But other than that, we don't really need `md5: 500f914dd06cf1f588b8e281c8fb3dd8` there, as we have param values anyways. So it should be something like:
```
deps:
  - path: params.conf
    params:
    - my_param: paramvaluefoundinfile
```
Might be a good idea to store params as a dict and not as a list, as identical param names don''t sound likely within the same conf file. But we might also want to extend that in the future with some special param markers, but that could work with both approaches (list of dicts or a key: dict), so we could start with either. Also I'm a bit worried about `my_param` name being compatible with yaml keys, but we could always move to `name: my_param` in dict format later.

**Comment by casperdcl** [comment]
I agree we shouldn't merge #3439 with this issue (even though in certain scenarios they may solve the same user requirement).

I am still super unconvinced by any plan to have DVC alter user files (regardless of whether it's with explicit consent and tackled 3 PRs from now). It's just plain wrong.

**Comment by elgehelge** [comment]
I few questions that comes to mind when implementing this:

Should we allow multiple references to the same file like `-d paramsfile -p paramsfile:foo`, or `-p paramsfile:foo -p paramsfile:bar` or `-p foo -p bar` or `-p a.* -p a.bar`? Currently I have just made non of these valid and instead implemented a slightly different syntax: `-p paramsfile:foo.*,bar,baz`. Just because it was easier, but let me know what you think, and I will adjust.

Did we decide on a name for the default params file? And the format?

**Comment by dmpetrov** [comment]
My suggestions:
1. **[must have]** A simple form for basic use case `-p LR -p epoch -p augm.angles`
2. **[nice to have]** The simple format can be easily extended to `-p LR,epoch,augm.angles`.
3. **[nice to have]** It would be great to support asterisk `-p augm.*` for groups of params. But it looks like it leads to an issue: you have to save the whole command with the asterisk in dvc-file as well as the expanded list of the params.
4. **[must have]** My preference for the default file name is `params.conf` (yaml, json are fine).
5. **[must have]** To overwrite the default file you can just introduce capital `-P process_params.yaml`. In most of the cases, only one param file will be used.
6. **[nice to have]** Granular file overwrite is good to have `-p LR -p epoch -p process_params.conf:augm.angles`

I'd suggest not to spend time on nice-to-have items in the first iteration. It can be easily extended later.

**Comment by dmpetrov** [comment]
I created a survey in a private data science community ods.ai

Q: What file format you use for conf file / hyper params files:
1. conf/ini
2. yaml
3. json
4. other (please clarify)

Results so far:
- conf/ini: 0
- **yaml: 12**
- json: 2
- other: 0

I'll update it in ~12 hours. But it seems like `params.yaml` a good default option.

**Comment by Suor** [comment]
I have a question: why we want this to be ini or json if we use yaml everywhere else?

On adding params to `.dvc` file, I would say it is a no go for a couple of reasons:
- those params are meant to affect more than one stage
- we are going to deprecate `.dvc` files for run commands in https://github.com/iterative/dvc/issues/1871, not sure when those will land though and design might still change
 
An alternative to a separate file is to add a section to new pipeline/"stage collection" file. See example [here](https://github.com/iterative/dvc/issues/1871#issuecomment-592184042) regarding metrics.

**Comment by casperdcl** [comment]
> Regarding `conf`/`ini`-vs-`json`, I'd say if wanting any level of manual user-editing `yaml` is best :)

It looks like #1871 is potentially blocking both #3439 and this (#3393), or at least confusing implementation. Maybe it would make more logical sense to tackle #1871 first?

**Comment by dmpetrov** [comment]
> On adding params to `.dvc` file

@Suor this `params.yaml` is not dvc-file, but just a params file. In dvc-files we need to save the subset of params that the stage depends on and values to identify changes.

**Comment by Suor** [comment]
> @Suor this params.yaml is not dvc-file, but just a params file. In dvc-files we need to save the subset of params that the stage depends on and values to identify changes.

So there won't be one param file, but lots of params spread around? Or this is like dep checksums?

**Comment by dmpetrov** [comment]
> Or this is like dep checksums?

Exactly. One params file `params.yaml` and dep-checkums spread around dvc-files (until single file is invented).

**Comment by Suor** [comment]
Then we have a conflict with collecting stages into single file again. As that file won't contain any checksums, so param values will also go to an accompanying `.lock` file and not be really visible. Is that an issue?

**Comment by dmpetrov** [comment]
> Is that an issue?

Not at all. However, I'd expect to have the params list in dvc-file, close to a stage definition but extract value (think checksums) to `.lock` file.

**Comment by Suor** [comment]
@dmpetrov adding values to the single stages file will bring issues. This will mean file is not entirely human controlled anymore. Which will result in:
- need to use slow yaml lib (`ruamel.yaml`) to handle that
- merge conflicts on generated data
- implementation will need to collect deps from two separate places: lock and stages file
- no way or at least harder to abstract out and switch "front-end", i.e. switch yaml stages files to python stages files

**Comment by dmpetrov** [comment]
> adding values to the single stages file will bring issues.

You are right. But I assume that `dvc run` still writes to dvc-file, keeps `*.lock` files and we have the same set of issues for that. I don't see how the params file makes it more complicated. I'd appreciate if you could clarify that.

**Comment by elgehelge** [comment]
I think we should tackle  #1871 first. Or at least not holde back with that. I will continue with this and make sure it is in a state where all the [must have]'s are satisfied, but I still don't think it should be merge before  #1871 is done and merged in. At that time we will have a better understanding of how these might conflict.

**Comment by elgehelge** [comment]
> 5\. **must have]** To overwrite the default file you can just introduce capital `-P process_params.yaml`. In most of the cases, only one param file will be used.

Instead of adding yet another option, maybe this is something that should be handled by an environment variable?

**Comment by dmpetrov** [comment]
> maybe this is something that should be handled by an environment variable?

Env variables break reproducibility.

**Comment by elgehelge** [comment]
> > maybe this is something that should be handled by an environment variable?
> 
> Env variables break reproducibility

Not regarding parameters since these are part of the dvc file.

As a side note, I would argue that environment variables should be supported as a dependency (but this is not a requirement for my argument to hold).

**Comment by elgehelge** [comment]
Just read up on MLFlow projects. In my opinion they have done it the right way.

```yaml
entry_points:
  main:
    parameters:
      data_file: path
      regularization: {type: float, default: 0.1}
    command: "python train.py -r {regularization} {data_file}"
  validate:
    parameters:
      data_file: path
    command: "python validate.py {data_file}"
```

A few great things about this approach:
* Not having to point at an external file
* Each parameter is closely coupled with the step (entrypoint) that uses the parameter
* They do not force a default value

**Comment by dmpetrov** [comment]
> Just read up on MLFlow projects. In my opinion they have done it the right way.

@elgehelge To my mind, this approach goes against the best practice of having a config file with hyper params. Probably 80%+ projects have config. It would be not easy to convince ppl to use this specific format. Also, I heard complaints about duplications in MlFlow pipelines - in your code example you repeated `regularization` twice, you need to repeat it again when you parse the args and one more time when you output it `mlflow.log_params("regularization", args.regularization)`.

In general, the MlFlow project seems not getting enough traction (comparing to MlFlow tracking) to use it as a good example.

**Comment by dmpetrov** [comment]
We did a great job by implementing the params #3515 (kudos to @elgehelge) and multi-stage dvc file which tracks params values through `dvc.lock` (thank you for @skshetry).

Closing this issue. Please LMK if there is something we need to extract to a separate one.
